File: go/133.clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

// 1. clone nodes and make a mapping from original nodes to copied nodes
// 2. connect edges
func cloneGraph(node *Node) *Node {
    if node == nil {
        return nil
    }
    // 1. BFS to traverse graph
    m := make(map[*Node]*Node)
    var queue, nextQueue []*Node
    queue = []*Node{node}
    for len(queue) > 0 {
        for len(queue) > 0 {
            from := queue[0]
            queue = queue[1:]
            if _, exists := m[from]; exists {
                continue
            }
            m[from] = &Node{Val: from.Val}
            nextQueue = append(nextQueue, from.Neighbors...)
        }
        queue, nextQueue = nextQueue, queue
    }
    
    // connect edges
    for original, copied := range m {
        for _, to := range original.Neighbors {
            copied.Neighbors = append(copied.Neighbors, m[to])
        }
    }
    
    return m[node]
}

File: csharp/133.clone-graph.cs
/*
 * @lc app=leetcode id=133 lang=csharp
 *
 * [133] Clone Graph
 */

// @lc code=start
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

using System.Collections.Generic;

public class Solution 
{
    public Dictionary<Node, Node> map = new Dictionary<Node, Node>();

    public Node CloneGraph(Node node) 
    {
        if (node == null)
        {
            return null;
        }

        Node newNode = null;

        if (!this.map.TryGetValue(node, out newNode))
        {
            newNode = new Node(node.val);
            this.map[node] = newNode;
            foreach (var neighbor in node.neighbors)
            {
                newNode.neighbors.Add(this.CloneGraph(neighbor));
            }
        }

        return newNode;
    }
}
// @lc code=end

